kernel/sched.c

```diff
@@ -1760,61 +1760,227 @@ static uint32_t usage_now(void)
 	return (now == 0) ? 1 : now;
 }
 
-void z_sched_usage_start(struct k_thread *thread)
+/**
+ * This routine currently assumes that [thread] runs on the current CPU.
+ */
+void z_sched_usage_start(struct k_thread *thread, bool new_window)
 {
-	/* One write through a volatile pointer doesn't require
-	 * synchronization as long as _usage() treats it as volatile
-	 * (we can't race with _stop() by design).
-	 */
-	_current_cpu->usage0 = usage_now();
```

Comment:
So... this is a **lot** heavier than the earlier implementation, which was expected to be like 4 cycles or whatever.  Remember that this is called out of the guts of the scheduler for every context switch.  We might be in interrupt context, we might be dealing with I/O bound, rapidly-switching threads, etc...

Is there no way to defer a bunch of this work until someone tries to look at the results via z_sched_cpu_usage() or whenever?

Reply:
Yeah, it is much heavier. I wasn't thrilled about that.

I've not yet been able to think of a way to move more of the work out of the context switch, but I will continue to ponder.